### ISSUE TITLE
[Question] How to compile MessageTemplate + Properties obtained from the HTTP sink?

### ISSUE THREAD

**Issue by ryanelian** [FantasticFiasco/serilog-sinks-http#5]
I'm trying to develop a logging server for my company using your HTTP sink, plugged into ASP.NET Core MVC.

The sink sends the following POST:

```json
{
  "events": [
    {
      "Timestamp": "2017-03-07T02:07:34.8287216+07:00",
      "Level": "Information",
      "MessageTemplate": "Executing ViewResult, running view at path {Path}.",
      "Properties": {
        "Path": "/Views/Home/Index.cshtml",
        "EventId": { "Id": 1 },
        "SourceContext": "Microsoft.AspNetCore.Mvc.ViewFeatures.Internal.ViewResultExecutor",
        "ActionId": "606648ce-b9fc-4811-8e5d-9d15b5e037c8",
        "ActionName": "Sample.Controllers.HomeController.Index (Sample)",
        "RequestId": "0HL34OG547AUO",
        "RequestPath": "/"
      }
    },
    {
      "Timestamp": "2017-03-07T02:07:34.8718259+07:00",
      "Level": "Information",
      "MessageTemplate": "User profile is available. Using '{FullName}' as key repository and Windows DPAPI to encrypt keys at rest.",
      "Properties": {
        "FullName": "C:\\Users\\Ryan\\AppData\\Local\\ASP.NET\\DataProtection-Keys",
        "SourceContext": "Microsoft.Extensions.DependencyInjection.DataProtectionServices",
        "ActionId": "606648ce-b9fc-4811-8e5d-9d15b5e037c8",
        "ActionName": "Sample.Controllers.HomeController.Index (Sample)",
        "RequestId": "0HL34OG547AUO",
        "RequestPath": "/"
      }
    },
    {
      "Timestamp": "2017-03-07T02:07:35.0372262+07:00",
      "Level": "Information",
      "MessageTemplate": "Executed action {ActionName} in {ElapsedMilliseconds}ms",
      "Properties": {
        "ActionName": "Sample.Controllers.HomeController.Index (Sample)",
        "ElapsedMilliseconds": 3271.4887000000003,
        "EventId": { "Id": 2 },
        "SourceContext": "Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker",
        "ActionId": "606648ce-b9fc-4811-8e5d-9d15b5e037c8",
        "RequestId": "0HL34OG547AUO",
        "RequestPath": "/"
      }
    },
    {
      "Timestamp": "2017-03-07T02:07:35.0467514+07:00",
      "Level": "Information",
      "MessageTemplate": "{HostingRequestFinished:l}",
      "Properties": {
        "ElapsedMilliseconds": 3598.5328,
        "StatusCode": 200,
        "ContentType": "text/html; charset=utf-8",
        "HostingRequestFinished": "Request finished in 3598.5328ms 200 text/html; charset=utf-8",
        "EventId": { "Id": 2 },
        "SourceContext": "Microsoft.AspNetCore.Hosting.Internal.WebHost",
        "RequestId": "0HL34OG547AUO",
        "RequestPath": "/"
      },
      "Renderings": {
        "HostingRequestFinished": [
          {
            "Format": "l",
            "Rendering": "Request finished in 3598.5328ms 200 text/html; charset=utf-8"
          }
        ]
      }
    }
  ]
}
```

Which is cool. However, I'm stuck with the process of converting the MessageTemplate + Properties into a compiled message.

Is there an instruction for doing so? Thank you in advance.

**Comment by FantasticFiasco** [comment]
Hi Ryan!

Before continuing, which version of the sink are you using? I just released v3.0.0, and chances are you're on an older version?

**Comment by ryanelian** [comment]
Hi!

These are my project dependencies and my setup.

```json
  "dependencies": {
    "HttpUtil": "0.6.0",
    "Microsoft.ApplicationInsights.AspNetCore": "2.0.0",
    "Microsoft.AspNetCore.Diagnostics": "1.1.0",
    "Microsoft.AspNetCore.Mvc": "1.1.1",
    "Microsoft.AspNetCore.Razor.Tools": {
      "version": "1.1.0-preview4-final",
      "type": "build"
    },
    "Microsoft.AspNetCore.Routing": "1.1.0",
    "Microsoft.AspNetCore.Server.IISIntegration": "1.1.0",
    "Microsoft.AspNetCore.Server.Kestrel": "1.1.0",
    "Microsoft.AspNetCore.Session": "1.1.0",
    "Microsoft.AspNetCore.StaticFiles": "1.1.0",
    "Microsoft.EntityFrameworkCore.SqlServer": "1.1.0",
    "Microsoft.Extensions.Caching.Redis": "1.1.0",
    "Microsoft.Extensions.Configuration.EnvironmentVariables": "1.1.0",
    "Microsoft.Extensions.Configuration.Json": "1.1.0",
    "Microsoft.Extensions.Options.ConfigurationExtensions": "1.1.0",
    "Microsoft.VisualStudio.Web.BrowserLink.Loader": "14.1.0",
    "Serilog.Extensions.Logging": "1.4.0",
    "Serilog.Sinks.Http": "3.0.0",
    "Serilog.Sinks.Literate": "2.1.0"
  },
```

```c#
Log.Logger = new LoggerConfiguration()
    .Enrich.FromLogContext()
    .WriteTo.Http(url, new Serilog.Sinks.Http.Options());
    .WriteTo.LiterateConsole()
    .CreateLogger();

loggerFactory.AddSerilog();
```

**Comment by FantasticFiasco** [comment]
My work leading up to release 3.0.0 involved replacing the deprecated JSON formatter with [the new and shining one](https://nblumhardt.com/2016/07/serilog-2-0-json-improvements/), however it seems that the new formatter doesn't render the message.

I can only guess that the folks at Serilog wanted to have the log events as compact as possible, and because of this dropped the rendered message property.

For this sink I can introduce a new options, making it possible for you to configure whether you wan't the message rendered or not. Would that work for you?

**Comment by ryanelian** [comment]
> For this sink I can introduce a new options, making it possible for you to configure whether you want the message rendered or not. Would that work for you?

That would be a welcomed addition! (Providing that the `Properties` aren't removed when the option is turned on because it may contain useful stuffs like `RequestId`, like the example above)

That said, my curiosity for how to render the message using the `MessageTemplate` + `Properties` is not yet sated. I feel like you don't need to add those additional options if I just knew how to render the message properly using C# / JavaScript.

**Comment by FantasticFiasco** [comment]
> That would be a welcomed addition! (Providing that the Properties aren't removed when the option is turned on because it may contain useful stuffs like RequestId, like the example above)

Of course, the properties would remain as before, the only change would be that the JSON payload now also has a rendered message. The rendered message would be the message template with the properties inserted at correct positions.

> That said, my curiosity for how to render the message using the MessageTemplate + Properties is not yet sated. I feel like you don't need to add those additional options if I just knew how to render the message properly using C# / JavaScript.

Am I correct in assuming that you would like to do this rendering on the receiving end of the log events, i.e. by the log server?

**Comment by ryanelian** [comment]
> Am I correct in assuming that you would like to do this rendering on the receiving end of the log events, i.e. by the log server?

Yes, exactly. Do you happen to know if Serilog has a method for doing that automatically?

**Comment by FantasticFiasco** [comment]
Yes they do, its `logEvent.MessageTemplate.Render(logEvent.Properties)`, you can see an example of it [here](https://github.com/serilog/serilog-formatting-compact/blob/dev/src/Serilog.Formatting.Compact/Formatting/Compact/RenderedCompactJsonFormatter.cs).

**Comment by ryanelian** [comment]
Thank you. I'll try it out sometimes.

**Comment by FantasticFiasco** [comment]
I've released [v3.1.0 of this sink](https://www.nuget.org/packages/Serilog.Sinks.Http/3.1.0), and I've addressed your issue about the missing rendered message.

I've added four different formatting types, where either `FormattingType.NormalRendered` or `FormattingType.CompactRendered` would be of interest for you. For a detailed explanation regarding the four types, please read the `README.md`.

If you feel happy with these changes, please close this issue and continue on enjoying Serilog and structured logging.

**Comment by ryanelian** [comment]
Spicy. 🌶 

Thank you for implementing and documenting this.
